arm-gcc.lua

```diff
@@ -0,0 +1,76 @@
+toolchain("arm-gcc")
+
+    set_description("ARM Compiler of GCC")
+    set_kind("cross")
+
+    on_load(function(toolchain)
+        if toolchain:is_plat("windows") then
+            toolchain:set("toolset", "cc", "arm-none-eabi-gcc.exe")
+            toolchain:set("toolset", "ld", "arm-none-eabi-gcc.exe")
+            toolchain:set("toolset", "ar", "arm-none-eabi-ar.exe")
+            toolchain:set("toolset", "as", "arm-none-eabi-gcc.exe")
+        else
+            toolchain:set("toolset", "cc", "arm-none-eabi-gcc")
+            toolchain:set("toolset", "ld", "arm-none-eabi-gcc")
+            toolchain:set("toolset", "ar", "arm-none-eabi-ar")
+            toolchain:set("toolset", "as", "arm-none-eabi-gcc")
+        end
+    end)
+
+toolchain_end()
+
+rule("generate-hex")
+    after_build(function(target)
+        local out = target:targetfile() or ""
+        local gen_fi = "build/" .. target:name()
+
+        -- https://github.com/xmake-io/xmake/discussions/2125
+        -- os.exec("arm-none-eabi-objdump -S "..out.." > "..gen_fi..".asm")
+        -- local asm = os.iorun("arm-none-eabi-objdump -S build/cross/cortex-m4/release/minimal-proj")
+        -- io.writefile(gen_fi..".asm", asm)
+
+        if target:is_plat("windows") then
+            os.execv("arm-none-eabi-objcopy.exe", {"-Obinary", out, gen_fi .. ".bin"})
+            os.execv("arm-none-eabi-objdump.exe", {"-S", out}, {stdout = gen_fi .. ".asm"})
+            os.execv("arm-none-eabi-objcopy.exe", {"-O", "ihex", out, gen_fi .. ".hex"})
+        else
+            os.execv("arm-none-eabi-objcopy", {"-Obinary", out, gen_fi .. ".bin"})
+            os.execv("arm-none-eabi-objdump", {"-S", out}, {stdout = gen_fi .. ".asm"})
+            os.execv("arm-none-eabi-objcopy", {"-O", "ihex", out, gen_fi .. ".hex"})
+        end
+
+        os.mv(out, gen_fi .. ".elf") -- add .elf file
+        --  -I binary
+        -- $(Q) $(OBJ_COPY) -O ihex $@ $(BUILD_DIR)/$(TARGET).hex
+        -- $(Q) $(OBJ_COPY) -O binary $@ $(BUILD_DIR)/$(TARGET).bin
+        -- os.exec("qemu-system-arm -M stm32-p103 -nographic -kernel"..bin_out)
+    end)
+
+    after_clean(function(target)
+        local gen_fi = "build/" .. target:name()
+        os.rm(gen_fi .. ".*")
+    end)
+
+rule_end()
+
+task("qemu")
+
+    on_run(function()
+        print("Run binary in Qemu!")
+        local bin_out = os.files("$(buildir)/*.bin")[1]
```

Comment:
use `import("lib.detect.find_file")` instead of os.files, and use `core.project.config` + config.buildir() instead of `$(buildir)`